### DISCUSSION TITLE
Question re: dependencies between client extensions

### DISCUSSION THREAD

**Discussion by scottyeck** [prisma/prisma#24276] [Q&A]
### Question

Can I reference a custom field added via a `result` extension in a `query` extension? I ask because...

* I haven't found anything in the docs that suggests otherwise (though I could be wrong)
* and (this is entirely subjective, but still notable!) intuitively the API for chaining extensions suggests that they are chained together as middlewares would be, and as such, the result from one would be passed to another.

My guess is that, despite my comment about chaining together the extensions, this is, in fact, not the case.

(To provide an additional layer of intrigue, this question was motivated by an upgrade from `v4` to `v5`, during which this was NOT a problem in `v4`, but IS in fact a problem in `v5`.)

### How to reproduce (optional)

1. Clone https://github.com/scottyeck/prisma-client-extension-playground
    * At time of writing, HEAD is [here](https://github.com/scottyeck/prisma-client-extension-playground/tree/4351b9a9603c0797eebd04c68085673bf813d1ce)
2. Run `yarn` to install.
3. Create a new Postgresql Database - add `DATABASE_URL` to `.env`
4. Run `npx prisma migrate deploy` to migrate schema
5. Run `npx tsx seed.ts` to seed data
6. Run `npx tsx index.ts`.

If you'd prefer to not repro and just read through the source code, here it is. It includes...

* The `kindExtension`, that adds a `kind` property to each model result.
* The `filterExtension`, that applies a custom filter in cases where the `kindExtension` indicates that the model is of kind `"user"`.
* The instantiation of a prisma client that uses these extensions (in the order they're described here).
* ...plus some code to actually invoke the thing.

```typescript
const filterExtension = Prisma.defineExtension((client) => {
  const extension = {
    query: {
      async $allOperations({ args, query }) {
        const result = await query(args);
        // TODO: this would need to consider one or many case, but for now
        // we'll just consider a many.
        if (result.length === 0) return result;
        return result.filter((item) => {
          if (item.kind === "user") {
            return item.name !== "Alice";
          }
          return true;
        });
      },
    },
  };
  return client.$extends(extension);
});

const kindExtension = Prisma.defineExtension((client) => {
  const filteredKeys = Object.keys(client).filter((key) => !["_", "$"].includes(key[0]));
  const result = filteredKeys.reduce((acc, key) => {
    acc[key] = {
      kind: {
        needs: {},
        compute() {
          return key;
        },
      },
    };
    return acc;
  }, {});
  const extension = { result };
  return client.$extends(extension);
});

const prisma = new PrismaClient().$extends(kindExtension).$extends(filterExtension);

async function main() {
  const allUsers = await prisma.user.findMany({
    include: {
      posts: true,
      profile: true,
    },
  });
  console.log(JSON.stringify(allUsers, null, 2));
}
```

### Expected behavior (optional)

As I mentioned, I would _expect_ that the result that's being filtered in the `filterExtension` would already have been fed through the `kindExtension` (because of the order in which the extensions have been configured). But as it turns out, that doesn't appear to be the case. The `kind` property isn't available yet in the `filterExtension`.

Is this expected? Is there something I've missed?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @scottyeck 

Thank you for raising this question.

You're correct about the functionality being available in v4 and no more in v5. The team made some changes to the behaviour in v5. An engineer will respond to this discussion with the reasoning behind this.

**Comment by SevInf** [comment]
Hi @scottyeck.
As @ludralph mentioned, this change was deliberate from our side and it happened here - https://github.com/prisma/prisma/pull/20438. One of the reasons for the change is described in linked issue #20437 - the whole classes of extensions we wanted to support - mocking, caching, routing, etc was not compatible with result extensions otherwise.  

Bigger problem is reliability of generic query extensions. Result extensions can change practically anything about the response and exposing that to the `query` extension makes it really difficult for a generic extension to rely on anything in the result: since you don't know what `result` extensions are applied together with your `query`, you have no clue of what you'd actually be getting if you call `query(args)`.

With that being said, you request totally makes sense - we are thinking of a more direct way of different extensions from the same bundle to communicate with each other and that would probably involve your query extension being able to read your own `result` extension results as well. Here is an issue to follow: https://github.com/prisma/prisma/issues/20788

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
